client/lib/media/actions.js

```diff
@@ -99,91 +100,139 @@ MediaActions.add = function( siteId, files ) {
+				date: new Date( baseTime - ( files.length - i ) ).toISOString()
+			};
 
-		Dispatcher.handleViewAction( {
-			type: 'CREATE_MEDIA_ITEM',
-			siteId: siteId,
-			data: transientMedia
-		} );
+			let createTransientMedia = () => {
+				if ( 'string' === typeof file ) {
+					assign( transientMedia, {
+						file: file,
+						extension: MediaUtils.getFileExtension( file ),
+						mime_type: MediaUtils.getMimeType( file ),
+						title: path.basename( file )
+					} );
+					return Promise.resolve();
+				}
+
+				assign( transientMedia, {
+					file: file.name,
+					extension: MediaUtils.getFileExtension( file.name ),
+					mime_type: MediaUtils.getMimeType( file.name ),
+					title: path.basename( file.name ),
+					// Size is not an API media property, though can be useful for
+					// validation purposes if known
+					size: file.size
+				} );
 
-		// Abort upload if file fails to pass validation.
-		if ( MediaValidationStore.getErrors( siteId, id ).length ) {
-			return Promise.resolve();
-		}
-
-		// Determine upload mechanism by object type
-		const isUrl = 'string' === typeof file;
-		const addHandler = isUrl ? 'addMediaUrls' : 'addMediaFiles';
-
-		// Assign parent ID if currently editing post
-		const post = PostEditStore.get();
-		if ( post && post.ID && ! isPlainObject( file ) ) {
-			file = {
-				parent_id: post.ID,
-				[ isUrl ? 'url' : 'file' ]: file
+				// If the image is a JPEG, fix the image orientation
+				if ( 'image/jpeg' === transientMedia.mime_type ) {
+					return ImageOrientationUtils.fixImageOrientation( file ).then( blobUrl => {
+						assign( transientMedia, {
+							URL: blobUrl,
+							guid: blobUrl
+						} );
+					} );
+				}
+
+				// Generate from window.File object
+				const fileUrl = window.URL.createObjectURL( file );
+				assign( transientMedia, {
+					URL: fileUrl,
+					guid: fileUrl
+				} );
+				return Promise.resolve();
 			};
-		}
-
-		debug( 'Uploading media to %d from %o', siteId, file );
-		return lastUpload.then( () => {
-			// Achieve series upload by waiting for the previous promise to
-			// resolve before starting this item's upload
-			const action = { type: 'RECEIVE_MEDIA_ITEM', id, siteId };
-			return wpcom.site( siteId )[ addHandler ]( {}, file ).then( ( data ) => {
-				Dispatcher.handleServerAction( Object.assign( action, {
-					data: data.media[ 0 ]
-				} ) );
-				// also refetch media limits
-				Dispatcher.handleServerAction( {
-					type: 'FETCH_MEDIA_LIMITS',
-					siteId: siteId
+
+			return createTransientMedia().then( () => {
+				Dispatcher.handleViewAction( {
+					type: 'CREATE_MEDIA_ITEM',
+					siteId: siteId,
+					data: transientMedia
+				} );
+
+				media.push( {
+					file: file,
+					transientMedia: transientMedia
 				} );
-			} ).catch( ( error ) => {
-				Dispatcher.handleServerAction( Object.assign( action, { error } ) );
 			} );
 		} );
-	}, Promise.resolve() );
+
+		return Promise.all( mediaPromises ).then( () => media );
```

Comment:
Rather than building an array via push, we can leverage the fact that `Promise.all` calls its thenable with an array of results from each promise, which is exactly how you've constructed this anyways.

e.g.
```js
Promise.all( [
	Promise.resolve( { ID: 1 } ),
	Promise.resolve( { ID: 2 } ),
	Promise.resolve( { ID: 3 } )
] ).then( ( media ) => { /* ... */ } );
```

So if you were to change `media.push( {} )` to `return {};`, you could drop the `.then()` here and eliminate the `media` variable.